File: 0735-asteroid-collision/0735-asteroid-collision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> st = new Stack<>();
        int sign = 0;
        for (int i = 0; i < asteroids.length; i++) {
            if (asteroids[i] > 0 || st.isEmpty() || st.peek() < 0)
                st.push(asteroids[i]);
            else if (st.peek() <= -asteroids[i])
                if (st.pop() < -asteroids[i]) i--;
        }
        return st.stream().mapToInt(i -> i).toArray();
    }
}

File: 0735-asteroid-collision/0735-asteroid-collision.kt
class Solution {
    fun asteroidCollision(asteroids: IntArray): IntArray {
       val stack = Stack<Int>()
        var i = 0
        while (i < asteroids.size) {
            if (asteroids[i] > 0 || stack.isEmpty() || stack.peek() < 0) stack.push(asteroids[i])
            else if (stack.peek() <= -asteroids[i]) {
                if (stack.peek() < -asteroids[i]) i--
                stack.pop()
            }
            i++
        }
        println(stack)
        val res = IntArray(stack.size)
        stack.fold(0) { acc, i ->
            res[acc] = i
            acc + 1
        }
        return res
    }
}